# 01 Importing libraries

In [3]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

# 02 Importing data

In [4]:
path = r'/Users/woodoooo/Desktop/Instacart Basket Analysis'

df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'),  index_col = False)

In [5]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [6]:
df_customers.shape

(206209, 10)

# 04 Wrangling the data -- renaming columns

In [7]:
df_customers.rename(columns = {'Surnam': 'last_name'}, inplace = True)

In [8]:
df_customers.rename(columns = {'STATE': 'state'}, inplace = True)

In [9]:
df_customers.rename(columns = {'Gender': 'gender'}, inplace = True)

In [10]:
df_customers.rename(columns = {'Age': 'age'}, inplace = True)

In [11]:
df_customers.rename(columns = {'First Name': 'first_name'}, inplace = True)

In [12]:
df_customers.rename(columns = {'fam_status': 'marital_status'}, inplace = True)

In [13]:
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants marital_status  income  
0             3        married  165665  
1             0         single   59285  
2             2        married   99568  
3             0         single   42049  
4             1        married   40374

# 05 Data quality and consistency checks

In [14]:
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype 
---  ------          --------------   ----- 
 0   user_id         206209 non-null  int64 
 1   first_name      194950 non-null  object
 2   last_name       206209 non-null  object
 3   gender          206209 non-null  object
 4   state           206209 non-null  object
 5   age             206209 non-null  int64 
 6   date_joined     206209 non-null  object
 7   n_dependants    206209 non-null  int64 
 8   marital_status  206209 non-null  object
 9   income          206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [15]:
# Changing the type of the 'date_joined' column to datetime type

df_customers['date_joined'] = pd.to_datetime(df_customers['date_joined'])

In [16]:
for col in df_customers.columns:
    # Check if there’s more than one type in the column
    if df_customers[col].apply(type).nunique() > 1:
        print(col)

first_name


In [17]:
# Change data types for multiple columns
df_customers = df_customers.astype({
    'user_id': 'int32',
    'age': 'int32',
    'n_dependants': 'int32',
     'income': 'int32'
    })

In [18]:
# Checking missing values 
df_customers['first_name'].isnull().sum()

np.int64(11259)

There are missing values in **first_name** (has **194,950 non‑null, so 11,259 missing)**

In [19]:
df_customers[df_customers['first_name'].isnull()].head(10)

user_id first_name   last_name  gender           state  age date_joined  \
53     76659        NaN     Gilbert    Male        Colorado   26  2017-01-01   
73     13738        NaN       Frost  Female       Louisiana   39  2017-01-01   
82     89996        NaN      Dawson  Female          Oregon   52  2017-01-01   
99     96166        NaN     Oconnor    Male        Oklahoma   51  2017-01-01   
105    29778        NaN      Dawson  Female            Utah   63  2017-01-01   
128     8562        NaN     Oconnor    Male            Utah   46  2017-01-01   
140   149267        NaN  Hutchinson    Male  South Carolina   20  2017-01-01   
149    82632        NaN         Orr    Male          Hawaii   61  2017-01-01   
155   172331        NaN  Williamson  Female          Alaska   27  2017-01-01   
236   182963        NaN   Nicholson  Female      New Mexico   58  2017-01-02   

     n_dependants marital_status  income  
53              2        married   41709  
73              0         single   82518  
82              3        married  117099  
99              1        married  155673  
105             3        married  151819  
128             1        married  134898  
140             0         single   86778  
149             1        married  118130  
155             0         single   55047  
236             1        married  163391

In [20]:
# dropping the first_name and last_name columns because there is no need to keep them in our analysis, and because of the privacy issues
df_customers.drop(columns=['first_name', 'last_name'], inplace=True)

In [21]:
df_customers.isnull().sum()

user_id           0
gender            0
state             0
age               0
date_joined       0
n_dependants      0
marital_status    0
income            0
dtype: int64

In [22]:
# Checking for duplicates 

df_dup = df_customers.duplicated()

In [23]:
df_dup.sum()

np.int64(0)

# 06 Importing the Instacart data  (4.8_ords_prods_merge_final.pkl)

In [24]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', '4.8_ords_prods_merge_final.pkl'))

In [25]:
df_ords_prods_merge.head()

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered  ... price_range_loc     Busiest day        Busiest_days  \
0          0  ...       Mid-range  Regularly busy  Regularly busy day   
1          0  ...       Mid-range  Regularly busy  Regularly busy day   
2          0  ...       Low-range  Regularly busy  Regularly busy day   
3          0  ...       Low-range  Regularly busy  Regularly busy day   
4          0  ...       Low-range  Regularly busy  Regularly busy day   

   busiest_period_of_day max_order  loyalty_flag avg_price avg_spending_flag  \
0         Average orders        10  New customer  6.367797       Low spender   
1         Average orders        10  New customer  6.367797       Low spender   
2         Average orders        10  New customer  6.367797       Low spender   
3         Average orders        10  New customer  6.367797       Low spender   
4         Average orders        10  New customer  6.367797       Low spender   

  user_median_order_frequency   order_frequency_flag  
0                        20.5  Non-frequent customer  
1                        20.5  Non-frequent customer  
2                        20.5  Non-frequent customer  
3                        20.5  Non-frequent customer  
4                        20.5  Non-frequent customer  

[5 rows x 25 columns]

In [26]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                       Dtype   
---  ------                       -----   
 0   order_id                     int32   
 1   user_id                      int32   
 2   order_number                 int8    
 3   order_day_of_week            int8    
 4   order_hour_of_day            int8    
 5   days_since_prior_order       float32 
 6   new_customer                 bool    
 7   product_id                   int32   
 8   add_to_cart_order            int32   
 9   reordered                    int8    
 10  product_name                 object  
 11  aisle_id                     int32   
 12  department_id                int32   
 13  prices                       float32 
 14  match                        category
 15  price_range_loc              object  
 16  Busiest day                  object  
 17  Busiest_days                 object  
 18  busiest_period_of_da

In [27]:
df_ords_prods_merge.shape

(32404859, 25)

In [28]:
df_ords_prods_merge['match'].head()

0    both
1    both
2    both
3    both
4    both
Name: match, dtype: category
Categories (3, object): ['left_only', 'right_only', 'both']

In [29]:
df_ords_prods_merge.drop(columns=['match'], inplace = True)

In [30]:
df_ords_prods_merge.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 24 columns):
 #   Column                       Dtype  
---  ------                       -----  
 0   order_id                     int32  
 1   user_id                      int32  
 2   order_number                 int8   
 3   order_day_of_week            int8   
 4   order_hour_of_day            int8   
 5   days_since_prior_order       float32
 6   new_customer                 bool   
 7   product_id                   int32  
 8   add_to_cart_order            int32  
 9   reordered                    int8   
 10  product_name                 object 
 11  aisle_id                     int32  
 12  department_id                int32  
 13  prices                       float32
 14  price_range_loc              object 
 15  Busiest day                  object 
 16  Busiest_days                 object 
 17  busiest_period_of_day        object 
 18  max_order                    int8   
 19

# 07 Merging the dataframes

In [31]:
# the common key is user_id

In [32]:
df_ords_prods_customers = pd.merge(df_ords_prods_merge,
                                  df_customers,
                                  on= 'user_id',  # common key column
                                  how = 'inner', # only rows that have a match in BOTH dataframes will be kept
                                  indicator=True)    # adds a column to show merge result

In [33]:
df_ords_prods_customers.shape

(32404859, 32)

In [34]:
df_ords_prods_customers.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   
5   2398795        1             2                  3                  7   
6   2398795        1             2                  3                  7   
7   2398795        1             2                  3                  7   
8   2398795        1             2                  3                  7   
9   2398795        1             2                  3                  7   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   
5                    15.0         False         196                  1   
6                    15.0         False       10258                  2   
7                    15.0         False       12427                  3   
8                    15.0         False       13176                  4   
9                    15.0         False       26088                  5   

   reordered  ... user_median_order_frequency   order_frequency_flag  gender  \
0          0  ...                        20.5  Non-frequent customer  Female   
1          0  ...                        20.5  Non-frequent customer  Female   
2          0  ...                        20.5  Non-frequent customer  Female   
3          0  ...                        20.5  Non-frequent customer  Female   
4          0  ...                        20.5  Non-frequent customer  Female   
5          1  ...                        20.5  Non-frequent customer  Female   
6          0  ...                        20.5  Non-frequent customer  Female   
7          1  ...                        20.5  Non-frequent customer  Female   
8          0  ...                        20.5  Non-frequent customer  Female   
9          1  ...                        20.5  Non-frequent customer  Female   

     state age date_joined n_dependants marital_status  income _merge  
0  Alabama  31  2019-02-17            3        married   40423   both  
1  Alabama  31  2019-02-17            3        married   40423   both  
2  Alabama  31  2019-02-17            3        married   40423   both  
3  Alabama  31  2019-02-17            3        married   40423   both  
4  Alabama  31  2019-02-17            3        married   40423   both  
5  Alabama  31  2019-02-17            3        married   40423   both  
6  Alabama  31  2019-02-17            3        married   40423   both  
7  Alabama  31  2019-02-17            3        married   40423   both  
8  Alabama  31  2019-02-17            3        married   40423   both  
9  Alabama  31  2019-02-17            3        married   40423   both  

[10 rows x 32 columns]

In [35]:
df_ords_prods_customers.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.240486e+07  3.240486e+07  3.240486e+07       3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01       2.738867e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00       1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01       3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01       6.000000e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01       2.090077e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110409e+01  2.559866e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   
std         4.246380e+00            8.378570e+00  1.408400e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   

             prices     max_order     avg_price  user_median_order_frequency  \
count  3.240486e+07  3.240486e+07  3.240486e+07                 3.240485e+07   
mean   7.790887e+00  3.305217e+01  7.790884e+00                 1.039776e+01   
min    1.000000e+00  1.000000e+00  1.000000e+00                 0.000000e+00   
25%    4.200000e+00  1.300000e+01  7.378478e+00                 6.000000e+00   
50%    7.400000e+00  2.600000e+01  7.811882e+00                 8.000000e+00   
75%    1.130000e+01  4.700000e+01  8.229228e+00                 1.300000e+01   
max    2.500000e+01  9.900000e+01  2.320000e+01                 3.000000e+01   
std    4.100591e+00  2.515525e+01  1.020555e+00                 7.032845e+00   

                age                    date_joined  n_dependants        income  
count  3.240486e+07                       32404859  3.240486e+07  3.240486e+07  
mean   4.946527e+01  2018-08-16 15:37:38.003034880  1.501896e+00  9.943773e+04  
min    1.800000e+01            2017-01-01 00:00:00  0.000000e+00  2.590300e+04  
25%    3.300000e+01            2017-10-22 00:00:00  1.000000e+00  6.700400e+04  
50%    4.900000e+01            2018-08-16 00:00:00  2.000000e+00  9.661800e+04  
75%    6.500000e+01            2019-06-09 00:00:00  3.000000e+00  1.279120e+05  
max    8.100000e+01            2020-04-01 00:00:00  3.000000e+00  5.939010e+05  
std    1.848558e+01                            NaN  1.118865e+00  4.305727e+04

In [36]:
df_ords_prods_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 32 columns):
 #   Column                       Dtype         
---  ------                       -----         
 0   order_id                     int32         
 1   user_id                      int32         
 2   order_number                 int8          
 3   order_day_of_week            int8          
 4   order_hour_of_day            int8          
 5   days_since_prior_order       float32       
 6   new_customer                 bool          
 7   product_id                   int32         
 8   add_to_cart_order            int32         
 9   reordered                    int8          
 10  product_name                 object        
 11  aisle_id                     int32         
 12  department_id                int32         
 13  prices                       float32       
 14  price_range_loc              object        
 15  Busiest day                  object        
 16

# 08 Exporting the data

In [38]:
os.makedirs(os.path.join(path, '02 Data', 'Prepared Data'), exist_ok=True)

In [39]:
df_ords_prods_customers.to_pickle(os.path.join(path, '02 Data','Prepared Data', '4.9_ords_prods_customers.pkl')) 